jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/udp/UdpHexDecoder.java

```diff
@@ -0,0 +1,245 @@
+        if (mapPayload == null)
+            return null;
+        OdeMapMetadata mapMetadata = new OdeMapMetadata(mapPayload);
+      
+        // Add header data for the decoding process
+        ZonedDateTime utc = ZonedDateTime.now(ZoneOffset.UTC);
+        String timestamp = utc.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"));
+        mapMetadata.setOdeReceivedAt(timestamp);
+
+        mapMetadata.setOriginIp(senderIp);
+        mapMetadata.setMapSource(MapSource.RSU);
+        mapMetadata.setRecordType(RecordType.mapTx);
+        mapMetadata.setRecordGeneratedBy(GeneratedBy.RSU);
+        mapMetadata.setSecurityResultCode(SecurityResultCode.success);
+
+        return JsonUtils.toJson(new OdeAsn1Data(mapMetadata, mapPayload), false);
+   }
+
+   public static String buildJsonSpatFromPacket(DatagramPacket packet){
+        String senderIp = packet.getAddress().getHostAddress();
+        int senderPort = packet.getPort();
+        logger.debug("Packet received from {}:{}", senderIp, senderPort);
+
+        // Create OdeMsgPayload and OdeLogMetadata objects and populate them
+        OdeAsn1Payload spatPayload = getPayloadHexString(packet, UperUtil.SupportedMessageTypes.SPAT);
+        if (spatPayload == null)
+            return null;
+        OdeSpatMetadata spatMetadata = new OdeSpatMetadata(spatPayload);
+
+        // Add header data for the decoding process
+        ZonedDateTime utc = ZonedDateTime.now(ZoneOffset.UTC);
+        String timestamp = utc.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"));
+        spatMetadata.setOdeReceivedAt(timestamp);
+
+        spatMetadata.setOriginIp(senderIp);
+        spatMetadata.setSpatSource(SpatSource.RSU);
+        spatMetadata.setRecordType(RecordType.spatTx);
+        spatMetadata.setRecordGeneratedBy(GeneratedBy.RSU);
+        spatMetadata.setSecurityResultCode(SecurityResultCode.success);
+
+
+        return JsonUtils.toJson(new OdeAsn1Data(spatMetadata, spatPayload), false); 
+   }
+
+   public static String buildJsonTimFromPacket(DatagramPacket packet){
+
+      String senderIp = packet.getAddress().getHostAddress();
+      int senderPort = packet.getPort();
+      logger.debug("Packet received from {}:{}", senderIp, senderPort);
+
+      // Create OdeMsgPayload and OdeLogMetadata objects and populate them
+      OdeAsn1Payload timPayload = getPayloadHexString(packet, UperUtil.SupportedMessageTypes.TIM);
+      if (timPayload == null)
+         return null;
+      OdeTimMetadata timMetadata = new OdeTimMetadata(timPayload);
+
+      // Add header data for the decoding process
+      ZonedDateTime utc = ZonedDateTime.now(ZoneOffset.UTC);
+      String timestamp = utc.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"));
+      timMetadata.setOdeReceivedAt(timestamp);
+
+      timMetadata.setOriginIp(senderIp);
+      timMetadata.setRecordType(RecordType.timMsg);
```

Comment:
Do we miss a TimSource here?

Reply:
It looks like there is not a Tim Source field defined in the OdeTimMetadata object. Given that all other message types have this we may want to consider adding it. However, doing so will result in a schema change that may break downstream components (Jackson deserialization doesn't accept extra fields by default). In either case, this PR matches the behavior of the existing logic, so adding that field might be out of scope for this PR.